File: 239 (Sliding Window Maximum)/cpp/Solution.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> dq;
        vector<int> res;

        for(int i =0 ;i<k-1;i++){
            while (!dq.empty() && nums[dq.back()]<nums[i]){
                dq.pop_back();
            }
            dq.push_back(i);
        }

        for(int i = k-1;i<nums.size();i++){
            while (!dq.empty() && nums[dq.back()]<nums[i]){
                dq.pop_back();
            }
            dq.push_back(i);

            if (i-k==dq.front()){
                dq.pop_front();
            }
            res.push_back(nums[dq.front()]);
        }
        return res;
    }
};

File: 239 (Sliding Window Maximum)/java/Solution.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> dq = new ArrayDeque<>();
        int[] res = new int[1+nums.length-k];
        int index = 0;

        for(int i = 0;i<k-1;i++){
            while (!dq.isEmpty() && (nums[dq.peekLast()]<nums[i])){
                dq.pollLast();
            }
            dq.offerLast(i);
        }

        
        for(int i = k-1;i<nums.length;i++){
            while (!dq.isEmpty() && (nums[dq.peekLast()]<nums[i])){
                dq.pollLast();
            }
            dq.offerLast(i);
            if (i-k==dq.peekFirst()) dq.pollFirst();

            res[index++] = nums[dq.peekFirst()];
        }
        return res;
    }
}